Command improvements for ergonomics and error handling (#3362)
rust-lang/rfcs: text/0000-command-ergonomics.md

and has become bogged down due to lack of decision
on the overall approach.  Some references:

 * [#84908](https://github.com/rust-lang/rust/issues/84908)
   Stabilisation tracking issue for `ExitStatusError` 
 * [#81452](https://github.com/rust-lang/rust/pull/81452)
   MR, closed as blocked: Add `#[must_use]` to `process::Command`, `process::Child` and `process::ExitStatus` 
 * [#89004](https://github.com/rust-lang/rust/pull/89004)
   MR, closed: Add convenience API to `std::process::Command`
 * [#88306](https://github.com/rust-lang/rust/pull/88306)
   MR, closed: `ErrorKind::ProcessFailed` and `impl From<ExitStatusError>`
 * [#93565](https://github.com/rust-lang/rust/pull/93565)
   MR, draft: `impl Try for ExitStatus`
 * [#73126](https://github.com/rust-lang/rust/issues/73126)
   issue: `std::process::Command` `output()` method error handling hazards
 * [#73131](https://github.com/rust-lang/rust/issues/73131)
   Overall tracking Issue for `std::process` error handling

## Currently-accepted wrong programs

The following incorrect program fragments are all accepted today and run to completion without returning any error:

> ```

Reviewer: ```suggestion
```rust
```

and has become bogged down due to lack of decision
on the overall approach.  Some references:

 * [#84908](https://github.com/rust-lang/rust/issues/84908)
   Stabilisation tracking issue for `ExitStatusError` 
 * [#81452](https://github.com/rust-lang/rust/pull/81452)
   MR, closed as blocked: Add `#[must_use]` to `process::Command`, `process::Child` and `process::ExitStatus` 
 * [#89004](https://github.com/rust-lang/rust/pull/89004)
   MR, closed: Add convenience API to `std::process::Command`
 * [#88306](https://github.com/rust-lang/rust/pull/88306)
   MR, closed: `ErrorKind::ProcessFailed` and `impl From<ExitStatusError>`
 * [#93565](https://github.com/rust-lang/rust/pull/93565)
   MR, draft: `impl Try for ExitStatus`
 * [#73126](https://github.com/rust-lang/rust/issues/73126)
   issue: `std::process::Command` `output()` method error handling hazards
 * [#73131](https://github.com/rust-lang/rust/issues/73131)
   Overall tracking Issue for `std::process` error handling

## Currently-accepted wrong programs

The following incorrect program fragments are all accepted today and run to completion without returning any error:

```rust
    Command::new("touch")

>         .args(&["/dev/enoent/touch-1"]);

Reviewer 2: ```suggestion
        .args(&["/nonexistent/touch-1"]);
```

(Rather than having to explain that `/dev/enoent` is a non-existent directory, and having people wonder if because it's under `/dev` it has some nonstandard semantics attached.)

    // ^ programmer surely wanted to actually run the command

    Command::new("touch")

>         .args(&["/dev/enoent/touch-2"])

Reviewer 2: ```suggestion
        .args(&["/nonexistent/touch-2"])
```

        .spawn()?;
    // ^ accidentally failed to wait, if programmer wanted to make a daemon
    //   or zombie or something they should have to write let _ =.

    Command::new("touch")

>         .args(&["/dev/enoent/touch-3"])

Reviewer 2: ```suggestion
        .args(&["/nonexistent/touch-3"])
```

to use the new methods (see Unresolved Questions).

## Use cases

We aim to serve well each of the following people:

 * Alice writes a CLI utility to orchestrate processes and wants top-notch error reporting from subprocesses.

 * Bob migrates ad-hoc automation scripts from bash to Rust.

 * Carol writes a generic application and just wants all errors to be maximally useful by default.

 * Dionysus wants to run `diff`, which exits `0` for "no difference",
   `1` for "difference found" and
   another value for failure.

(Partly cribbed from one of
@matklad's [comments](https://github.com/rust-lang/rust/pull/89004#issuecomment-923803209) in #89004)

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

## New methods on `Command`:

 * `fn run(&mut self) -> Result<(), SubprocessError>`:

   Runs the command.
   Equivalent to `.spawn()` followed by `.status()`,
   but with better error handling.

 * `fn get_output_bytes(&mut self) -> Result<Vec<u8>, SubprocessError>`:

   Runs the command and collects its stdout.
   After the child indicates EOF on its stdout,
   we will wait for it to finish and check the exit status.

>  * `fn get_output(&mut self) -> Result<Vec<String>, SubprocessError>`:

Reviewer 2: ```suggestion
 * `fn get_output(&mut self) -> Result<String, SubprocessError>`:
```

I'm not sure if this was a typo, or if you intended this to have built-in split-by-lines. But in any case, we have methods for splitting by lines, so I think this should return the whole output as one `String`.

Author: This was a mistake, thanks.

   Runs the command and collects its stdout.
   Decodes the stdout as UTF-8, and fails if that's not possible.
   Does not trim any trailing line ending.

>  * `fn get_output_line(&mut self) -> Result<Vec<String>, SubprocessError>`:

Reviewer 2: ```suggestion
 * `fn get_output_line(&mut self) -> Result<String, SubprocessError>`:
```

This being `Vec<String>` as well is what makes me think the above is a typo, since otherwise, this Vec would always contain one element.

   Runs the command and collects its stdout.
   Decodes the stdout as UTF-8, and fails if that's not possible.

>    Fails unless the output is a single line (with or without line ending).

Reviewer 2: Please specify the behavior if the program has *no* output. Does this fail in that case, or is the requirement "at most one line"?

Author: I think it should succeed and give an empty string.  Will write that into the RFC.

Reviewer 2: It seems like this helper would commonly be used for programs for which you expect to parse the output, in which case having to separately check for empty string seems to defeat part of the ergonomic win of using the helper.

But in any case, this isn't a blocker as you've split this helper out into possible future work.

   Trims the line ending (if any).

 * `fn get_output_read(&mut self) -> std::process::ChildOutputStream`
   (where `struct ChildOutputStream` implements `io::Read`
   and is `Send + Sync + 'static`).

   Starts the command, allowing the caller to
   read the stdout in a streaming way.
   Neither EOF nor an error will be reported by `ChildOutputStream`
   until the child has exited, *and* the stdout pipe reports EOF.
   (This includes errors due to nonempty stderr,

>    if stderr was set to `piped`.)

Reviewer 2: This description makes me wonder about the semantics of the *other* functions in this family. Do they wait for process exit, or for process exit *and* EOF? (I'm hoping the former, or possibly "process exit and then a non-blocking read saying it would block".)

Author: This is stated explicitly in the docs for what I am now calling `read_stdout_bytes`.  All of these functions wait for both EOF and process exit.  I will add a short xref note to the other two functions.  In the real docs the spec text should probably be cut and pasted instead.

Most callers should use these methods,
rather than `.spawn()` or `.status()`.
These new methods should be be
recommended by docs for other, lower-level functions
which people were previously required to use.

## Deprecations

>  * Deprecate `std::process::Command::output()`.

Reviewer 3: It's not clear to me that `Command::output` is that bad. It's slightly awkward and requires manually checking errors, but I'm not convinced this is worth deprecating a *extremely* widely-used function and causing a large amount of ecosystem churn (which many users will respond to with `#[allow(deprecated)]`, due to MSRV+the existing method not being broken).

I think you need to make a much clearer case for why this needs deprecation, and document the downside of ecosystem and educational churn as folks move to the new API(s).

Author: I think you and Josh are right about this.  I have moved the deprecation to Future possibilities.

   This API cannot be fixed;
   see [#73126](https://github.com/rust-lang/rust/issues/73126).

>  * Apply `#[must_use]` to `Command` and `Child`.

Reviewer 2: :tada: :tada: :tada:

Reviewer 4: Would it be possible to make this change without an RFC? Seems somewhat unfortunate to tie this together with the API additions.

but it may represent several

For example a process which exited nonzero
probably printed to stderr;
with `piped` we capture that, and represent both the stderr
and the exit status as problems within the `SubprocessError`.

```rust
impl SubprocessError {
    /// The program, if we know it.
    fn program(&self) -> Option<&OsStr>;

    /// The arguments, if we know them.
    fn args(&self) -> Option<std::process::CommandArgs<'_>>;

    /// If the stdout was captured in memory, the stdout data.
    fn stdout_bytes(&self) -> Option<&[u8]>;

    /// If the process exited and we collected its status, the exit status.
    fn status(&self) -> Option<ExitStatus>;

    /// If trouble included nonempty stderr, the captured stderr
    fn stderr_bytes(&self) -> Option<&[u8]>;

    /// If trouble included failure to spawn, the spawn error.
    fn spawn_error() -> Option<&io::Error>;

    /// If trouble included failure to talk to the child, the IO error.
    ///
    /// This might include problems which might be caused by child
    /// misbehaviour.
    fn communication_error() -> Option<&io::Error>;

    /// If trouble included failed UTF-8 conversion.

>     fn utf8_error(&self) -> Option<&std::str::FromUtf8Error>;

Reviewer 3: Which UTF-8 conversion does this refer to? I think this probably should be pushed down to whatever getter would return something UTF-8 validated rather than having this stored as a field on the error.

If this *is* stored (and almost certainly if it is not), we should use `std::std::Utf8Error` rather than a `&std::string::FromUtf8Error` in order to avoid needing to store the invalid UTF8 twice (and to avoid forcing us to convert eagerly).

Author: I'm using `std::str::FromUtf8Error` which doesn't contain or reference the string, but does contain the offset.  I think this doesn't suffer from the problem you describe?

  * The `io::ErrorKind` from the spawn error, if any.

  * Otherwise, a new kind `io::ErrorKind::ProcessFailed`,
    which means that the subprocess itself failed.

### Further necessary APIs for `SubprocessError`

We also provide ways for this new error to be constructed,
which will be needed by other lower level libraries besides std,
notably async frameworks:

```rust
impl SubprocessError {
    /// Makes a "blank" error which doesn't contain any useful information
    ///
    /// `has_problem()` will return `false` until one of the setters
    /// is used to store an actual problem.
    fn new_empty() -> Self { }

    // If we keep ExitStatusError
    fn from_exit_status_error(status: ExitStatusError) -> Self { }

    fn set_program(&mut self, impl Into<OsString>);
    fn set_args(&mut self, impl IntoIterator<Item=impl Into<OsString>>);
    fn set_stdout_bytes(output: Option<Into<Box<[u8]>>>);

    fn set_status(&mut self, status: ExitStatus);
    fn set_stderr_bytes(&mut self, stderr: impl Into<Box<u8>>);
    fn set_spawn_error(&mut self, error: Option<io::Error>);
    fn set_communication_error(&mut self, error: Option<io::Error>);

>     fn set_utf8_error(&mut self, error: Option<std::str::FromUtf8Error>);

Reviewer 3: I think too much of this is mutable -- either we should just make these be public fields (we have `get_` and `set_` methods for them...) or we should just allow constructing one of these from parts.

Additionally it's not clear to me how these interact. Up until now, I had imagined that there was an internal enumeration 

```rs
enum SubprocessErrorCause {
    Spawn(io::Error),
    Communication(io::Error),
    // Actually, I don't think we want this one, but anyway
    Utf8(str::Utf8Error),
    // ...
}
```
but this API has individual setters for each one, which implies it either is not that way, or they have perhaps-confusing interactions.

Does it make any sense for there to be more than one error here anway? Which would gets reported in `std::error::Error::source`?

Author: I have made this a transparent struct.  Making it opaque was indeed confusing.

    /// Find out if this error contains any actual error information
    ///
    /// Returns `false` for a fresh blank error,
    /// or `true` for one which has any of the error fields set.
    /// (Currently equivalent to checking all of `status()`,
    /// `stderr_bytes()`, `spawn_error()` and `utf8_error()`.)
    //
    // We must provide this because it's needed for handling programs
    // with unusual exit status conventions (eg `diff(1)`)
    // and a caller can't reimplement it without making assumptons
    // about `SubprocessError`'s contents.
    fn has_problem(&self) -> bool;
}
impl Default for SubprocessError { ... }

> impl Clone for SubprocessError { ... } // contained io:Errors are in Arcs

Reviewer 3: I think we should instead ask users who need a `Clone`able `SubprocessError` to use `Arc<SubprocessError>`, rather than imposing that requirement on the internals.

Author: Fair enough, I have changed this.

[prior-art]: #prior-art

Many other languages have richer or more convenient APIs
for process invocation and output handling.

 * Perl's backquote operator has the command inherit the script's stderr. If you want to do something else you need to do a lot of hand-coding. It reports errors in a funky and not particularly convenient way (but, frankly, that is typical for Perl). It doesn't trim a final newline but Perl has a chomp operator that does that very conveniently.

 * Tcl's `exec` captures stderr by default (there are facilities for redirecting or inheriting it), calling any stderr an error (throwing a Tcl exception). It always calls nonzero exit status an error (and there is no easy way to get the stdout separately from the error in that situation). It unconditionally chomps a final newline.

 * Python3's `subprocess.run` fails to call nonzero exit status an exception. If you don't write explicit error handling code (highly unusual in Python) you have an unchecked exit status bug.

# Unresolved questions
[unresolved-questions]: #unresolved-questions

## Printing command arguments in `impl Display for SubprocessError`

Perhaps printing the command arguments is overly verbose,

> and we should print onliy the command name.

Reviewer 2: ```suggestion
and we should print only the command name.
```

## `get_output` vs `read_output` naming.

We have `fs::read_to_string`.

Possible names (taking `get_output_bytes` as the example):

 * `get_output_bytes` (proposed in this RFC)
 * `output_bytes` but `output` is alread taken for bad `Output`.
 * `run_get_output_bytes`
 * `run_output_bytes`
 * `read_output_bytes`

## stderr handling default

What should happen if the Rust programmer didn't call `.stderr()` ?

Options are:

 1. Treat it as `inherit`.
   This is probably fine for a command line application,
   but may be too lax for many other contexts.

 2. Treat it as `piped`: call any stderr output an error.
   This is a fairly conservative choice,
   but it can lead to unexpected runtime errors,
   if a called program later starts printing warning messages.

 3. Call this a programming mistake, and panic.

 4. Somehow make this a compile error.
   This would involve a new `CommandForWhichWeHaveSpecifiedStderr`
   type (typestate pattern), or providing the stderr handling as a mandatory argument
   to all the new methods.
   These seem unpalatably unergonomic.

> 

Reviewer 2: ```suggestion

5. One of the above, and additionally provide a lint that makes a best-effort attempt to catch this at compile time.

```

    /// misbehaviour.
    pub communication_error: Option<io::Error>,

    /// If we had a problem converting stdout to UTF-8.
    ///
    /// The `error_len()` and `valid_up_to()` reference positions in `stdout_bytes`.
    pub utf8_error: Option<std::str::FromUtf8Error>,
}
impl Debug for ProcessError {
    // print all the fields except `stdout_bytes`.
}
impl Error for ProcessError {
    // fn cause() always returns None.
}
```

The `Display` implementation will print everything above
including the command's arguments.
The arguments will be escaped or quoted in some way that renders 
a resulting error message unambiguous.

### `impl From<ProcessError> for io::Error`

`ProcessError` must be convertible to `io::Error`
so that we can use it in `ChildOutputReader`'s
`Read` implementation.
This may also be convenient elsewhere.

The `io::ErrorKind` for a `ProcessError` will be:

  * The `io::ErrorKind` from the spawn error, if any.

  * Otherwise, a new kind `io::ErrorKind::ProcessFailed`,
    which means that the subprocess itself failed.

### Further APIs for `ProcessError`

A `ProcessError` is a transparent `Default` struct so it can be

> constructed outside std, for example by async frameworks.

Reviewer 2: ```suggestion
constructed outside std, such as by async frameworks or other
code launching processes and handling errors from them.
```

 * Tcl's `exec` captures stderr by default (there are facilities for redirecting or inheriting it), calling any stderr an error (throwing a Tcl exception). It always calls nonzero exit status an error (and there is no easy way to get the stdout separately from the error in that situation). It unconditionally chomps a final newline.

 * Python3's `subprocess.run` fails to call nonzero exit status an exception. If you don't write explicit error handling code (highly unusual in Python) you have an unchecked exit status bug.

# Unresolved questions
[unresolved-questions]: #unresolved-questions

## Printing command arguments in `impl Display for ProcessError`

Perhaps printing the command arguments is overly verbose,
and we should print only the command name.

Sometimes people pass sensitive information (such ass passwords) in command line arguments.

> This is not a good idea, because command line arguments are generally public on Unix.

Reviewer 2: ```suggestion
This is not a good idea in portable software, because command line arguments are often public on Unix targets.
```

   type (typestate pattern), or providing the stderr handling as a mandatory argument
   to all the new methods.
   These seem unpalatably unergonomic.

5. One of the above, and additionally provide a lint that makes a best-effort attempt to catch this at compile time.

Here we propose option 1: treat as `inherit`.

# Examples

## Conveniently running `diff`

```rust
    let result = Command::new("diff")
        .args(["before","after"])
        .stderr(Stdio::piped()) // optional, could just let it inherit
        .run();
    let status = match result {
        Ok(()) => 0,
        Err(err) => err.just_status()?.code(),
    };
```

# Future possibilities
[future-possibilities]: #future-possibilities

## Providing a way to combine and interleave stdout and stderr

Currently, `Command` insists on separating out stdout and stderr,
if you ask to capture them.
If you want them combined
(which is the only way to preserve the relative ordering)
you must do one of:

 * run a command which itself does the redirection
   (easy using the shell on Unix)

 * send them each to your own stdout/stderr with `inherit`
   and expect your caller to combine them

>  * sent themk to the *same* one of your stdout/stderr

Reviewer 2: ```suggestion
 * send them to the *same* one of your stdout/stderr
```

   which will be possible after
   https://github.com/rust-lang/rust/pull/88561)

We should provide something like this:

```
impl Command {
    /// Arranges that the command's stderr will be sent to wherever its stdout is going
    fn stderr_to_stdout();
}
```

> (It is not sensibly possible at least on Unix

Reviewer 2: ```suggestion
(It is not easy on Unix targets
```

It is possible, though quite awkward and with some limitations. (Even more limitations if you're not on Linux.) I needed a solution to this, and ended up writing [io-mux](https://crates.io/crates/io-mux) for this.

   and is `Send + Sync + 'static`).

   Starts the command, allowing the caller to
   read the stdout in a streaming way.
   Neither EOF nor an error will be reported by `ChildOutputReader`
   until the child has exited, *and* the stdout pipe reports EOF.
   (This includes errors due to nonempty stderr,
   if stderr was set to `piped`.)

Most callers should use these methods,
rather than `.spawn()` or `.status()`.
These new methods should be be
recommended by docs for other, lower-level functions
which people were previously required to use.

## Deprecations

 * Apply `#[must_use]` to `Command` and `Child`.

 * Apply `#[must_use]` to `ExitStatus`.
   (May require fixing quite a few of the examples.)

 * Add a warning to the docs for `Command.output()` about the lost error bugs,
   in particular the need to check `.status` and the lack of any compiler
   warning if one doesn't.  Suggest to the reader to consider
   `.run()` or `.read_stdout*` instead.
   Do not deprecate `.output()` though.

## stderr handling

If `stderr(Stdio::piped())`,
these new functions all collect the child's stderr.
Then,
if the stderr output is nonempty, this is considered an error,

> and reported in the `ProcessError`.

Reviewer 5: I'm a bit uncertain about this semantics.  It's popular in command-line applications to use `stderr` to collect running commentary about what the program is doing (for example `log::debug` under `env_logger`) in ways that definitely shouldn't be considered errors.

Reviewer 6: I have an experience report that erroring out on non-empty stderr is wrong! One common case here is when you run something like `rustc --version`, but that thing is `rustup` managed, so, before you get to `rustc --version` output, you have printout from `rustup` on the stderr that it downloads the appropriate version.

    /// Otherwise returns `self`.
    ////
    /// Use this if you want to to tolerate some exit statuses,
    /// but still fail if there were other problems.
    //
    // This is optional, and could be a separate feature from the rest of the RFC.
    // But it does make running programs like `diff` considerably easier.
    // (It is also implementable externally in terms of .has_problem()`.)
    pub fn just_status(self) -> Result<ExitStatus, ProcessError>;
}
impl Default for ProcessError { ... }
```

# Drawbacks
[drawbacks]: #drawbacks

This is nontrivial new API surface.

Much of the new API surface is in `ProcessError`.
If we didn't want to try to make it easy for Rust programmers
to run subprocesses and produce good error messages,
we could omit this error type 
(perhaps using something like `ExitStatusError`).

Perhaps we don't need all the `read_stdout` variants,
and could require Bob to write out the boilerplate
or provide his own helper function.

Perhaps we don't need `stdout_reader`.
However,
avoiding deadlocks when reading subprocess output,
and also doing error checks properly,

> is rather subtle.

Reviewer 2: Could you clarify how `stdout_reader` makes avoiding this deadlock easier?

Is this just that you need to interleave `.write` calls to the process stdin with `.read` calls to the reader, rather than calling `.write_all` and then reading all the output at once?

Author: The deadlocks that `stdout_reader` avoids completely are ones ike: waiting for the command's stdout EOF, and child termination, in the wrong order; waiting for stderr EOF (which can sometimes block forever for complicated reasons).

It helps only a little with deadlocks resulting from piping both into and out of the same `Command`.  (I think it may still help in those use cases, but it *doesn't* help with "you wrote too much to the command's input while it's trying to write output to you".)

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

> Alternatives and prior proposals include:

Reviewer 7: Remark: one other possible alternative: don't introduce these APIs yet, and try to cook up some kind of wrapper on top of `std::process::Command` outside std first. It seems to me that different users may have different preferences for the `std::process::Command` APIs (especially considering how hard it is to reach any kind of consensus), which pulls the API design in many directions.

for process invocation and output handling.

 * Perl's backquote operator has the command inherit the script's stderr. If you want to do something else you need to do a lot of hand-coding. It reports errors in a funky and not particularly convenient way (but, frankly, that is typical for Perl). It doesn't trim a final newline but Perl has a chomp operator that does that very conveniently.

 * Tcl's `exec` captures stderr by default (there are facilities for redirecting or inheriting it), calling any stderr an error (throwing a Tcl exception). It always calls nonzero exit status an error (and there is no easy way to get the stdout separately from the error in that situation). It unconditionally chomps a final newline.

 * Python3's `subprocess.run` fails to call nonzero exit status an exception. If you don't write explicit error handling code (highly unusual in Python) you have an unchecked exit status bug.

# Unresolved questions
[unresolved-questions]: #unresolved-questions

## Printing command arguments in `impl Display for ProcessError`

Perhaps printing the command arguments is overly verbose,
and we should print only the command name.

> Sometimes people pass sensitive information (such ass passwords) in command line arguments.

Reviewer 7: `ass` -> `as` (sorry)

    // and a caller can't reimplement it without making assumptons
    // about `SubprocessError`'s contents.
    fn has_problem(&self) -> bool;

    /// Returns `Ok<ExitStatus>` if the only reason for the failure was a nonzero exit status.
    /// Otherwise returns `self`.
    ////
    /// Use this if you want to to tolerate some exit statuses,
    /// but still fail if there were other problems.

>     pub fn just_status(self) -> Result<ExitStatus, SubprocessError>;

Reviewer 2: This feels a little awkward to use: if you call this to filter out some exit statuses, you lose the rest of the error, which means you don't have `stdout_bytes` anymore. What code pattern would you suggest for code that wants to run `diff`, check the exit code for 0/1 to determine if there was a difference, bail on >1, and capture the patch output?

Reviewer 2: I wonder if this (straw proposal) might be easier to use:

```rust
/// If the process is only considered to have errored due
/// to its exit code, calls the provided filter, returning
/// `Ok(Self)` if it returns `true`, or `Err(Self)` if it
/// returns `false`.
pub fn permit_statuses(self, f: impl FnOnce(ExitStatus) -> bool) -> Result<ProcessError, ProcessError>
```

This would make it relatively straightforward to say "allow 0 or 1, reject everything else". (Still mildly annoying due to `code()` returning `Option`, but doable.)

Author: I quite like this idea.  I wasn't sure people would like the notion of returning `Ok(ProcessError)`, but if you think that's OK it seems OK to me.

>  mildly annoying due to `code()` returning `Option`

I wondered whether the closure should receive `ExitCode` (so would only be called if the process exited rather than was killed).  But then you can't use this to ignore (say) `SIGPIPE`.

I do kind of hate booleans.  Would `impl FnOnce(ExitStatus) -> Result<(), ()>` be too awful?

Perhaps printing the command arguments is overly verbose,
and we should print only the command name.

## `cause` in `Error` impl of `ProcessError`

Because maybe several things went wrong, ever providing a `Some` `cause`

> would involve prioritising multiple possible problems.

Reviewer 2: Can you give an example where multiple different things could simultaneously go wrong? It *seems* like `spawn_error`, `communication_error`, and `utf8_error` should be mutually exclusive.

Author: I think the following can all occur in the same execution:
 * The application calls `.stderr(piped)` and `.read_stdout()`
 * Spawn succeeds
 * The subprocess produces some invalid-UTF output
 * Then, something in the kernel goes very badly awry (or, some other thing in our Rust program wrecks our fd) and `read()` on one of the pipes fails
 * The subprocess prints some stderr (which the application has said to treat as an error)
 * The subprocess then crashes.

We must report *at least* all of: the error from `read`; the stderr output (which might well be very helpful for diagnosis); and the process's exit status.

I think we should eagerly do a unicode validity check whenever we captured the stdout, and record the unicode conversion error in the `ProcessError`.  That way a caller who calls your `.permit_status()` and gets `Ok` can know that the stdout output is UTF-8; conversely, if the command unexpectedly printed non-UTF-8, the caller can conveniently report the command line etc.

Or to put it another way, we shouldn't "short circuit" some checks and report only a subset of the errors.

}
```

> (It [can be difficult or impossible](https://docs.rs/io-mux/latest/io_mux/) at least on Unix

Reviewer 2: (I don't actually need this cited, it's off-topic for this RFC, I just didn't want it described as impossible. :) )

Author: I know you weren't wanting a plug but I like to give references :-).  If I had been the reader rather than the author, I would have appreciated that link which is why I put it in.  I can take it out if it's not appropriate...

Reviewer 2: Perfectly fine, I just wanted to make sure it hadn't come across like I was asking for a plug.
